### PR TITLE
Progress bar context handler and decorator

### PR DESCRIPTION
**Pull request by janezd**
Packs `progressBarInit` and `progressBarFinished` into context manager to ensure the progress bar is removed in case of exceptions.

If PR is accepted, this should become the preferred way of handling progress bars.

### CODE REVIEW

File: Orange/widgets/widget.py [hunk 1]
```diff
@@ -539,6 +539,48 @@ def progressBarFinished(self, processEvents=QEventLoop.AllEvents):
         if processEvents is not None and processEvents is not False:
             qApp.processEvents(processEvents)
 
+    def progressBar(self, show_bar=True):
```
Comment (kernc) [inline Orange/widgets/widget.py]: What is `show_bar` and why would it ever be set to `False`?

Comment (kernc) [inline Orange/widgets/widget.py]: The current _preferred_ progress bar interface is to use `gui.progressBar` which provides a nice `advance()` wrapper for when the number of steps is known in advance.

Could this method accept instead a number of steps (like the one in `gui` does) and expose a similar method on the in-`__enter__`-returned `ProgressBarHandler` object?

File: Orange/widgets/widget.py [hunk 2]
```diff
@@ -539,6 +539,48 @@ def progressBarFinished(self, processEvents=QEventLoop.AllEvents):
...
+        Decorator and context manager for progress bar.
+
+        Using this decorator ensures that the progress bar is removed at the
+        end of the with-block or at exit from function even in case of
+        exceptions, therefore removing the need to remove the bar in
+        `finally` blocks.
+
+        Usage as context handler:
+
+            with self.progressBar:
```
Comment (kernc) [inline Orange/widgets/widget.py]: The examples above make a _call_ to `self.progressBar()`.

File: Orange/widgets/widget.py [hunk 3]
```diff
@@ -539,6 +539,48 @@ def progressBarFinished(self, processEvents=QEventLoop.AllEvents):
...
+                do_something()
+                update_progress_bar_while_doing_it()
+
+        Usage as decorator:
+
+            @OWWidget.progressBar
+            def compute(self):
+                do_something()
+                update_progress_bar_while_doing_it()
```
Comment (kernc) [inline Orange/widgets/widget.py]: I don't like this usage. It removes the progress bar initialization logic away from the code and into a decorator. This implies that either the code with a call to `progressBarSet()` (the example in this docstring could be clearer in this regard — what's `update_progress_bar_while_doing_it()`?), or the decorator could both be used on their own, which is obviously not the case.

There should be one obvious way to do it. Nothing wrong with `with self.progressBar()` (other than it increments indent level, but so does `finally`).

**Comment by kernc** [comment]
If this should become the preferred way, would you port existing code in widgets (i.e. make more examples) and mark the `gui.ProgressBar` [deprecated](https://github.com/biolab/orange3/blob/master/Orange/util.py#L14).

**Comment by janezd** [comment]
@kernc Now that I reimplemented it, I totally agree: decorators were a terrible idea. Context managers are nice, and the argument with the number of iterations simplifies the existing code. Please review the PR.

I haven't touched the old interface. There are some cases in which the init, advance and finish appear in different functions, therefore we will keep the old interface as a low-level functions for progress bars. For the same reason, `gui.ProgressBar` still makes sense.

We now have three mechanisms: context handler, widget's methods and `gui.ProgressBar`. They are useful for different cases, though.

File: Orange/widgets/widget.py [hunk 4]
```diff
@@ -570,6 +570,39 @@ def progressBarFinished(self, processEvents=QEventLoop.AllEvents):
...
+
+            with self.progressBar(20) as progress:
+                ...
+                progress.advance()
+
+        or
+
+            with self.progressBar() as progress:
+                ...
+                progress.advance(15)
```
Comment (kernc) [inline Orange/widgets/widget.py]: Calling `advance()` without initializing with total number of iterations results in what? 1500%?

**Comment by kernc** [comment]
If you fix the conflicts #889 introduced, this looks good to me.

Did you get to replace all the uses of `progressBarInit()` (& co.) in the library?

**Comment by janezd** [comment]
I resolved the conflicts.

I also replaced all uses of `progressBarInit()` wherever possible, that is, wherever the init and tear down were in the same block.

**Comment by kernc** [comment]
Would you like to [squash these two commits](https://www.google.com/search?q=github+squash+commits) since the first one includes just about everything and the second one is only a single-line fix-up?

**Comment by kernc** [comment]
Can you please rebase this once again so we can merge?
Thanks. :smile:

**Comment by janezd** [comment]
Rebased.

**Pull request merged**